File: python/elements/replicable_graph_element.py
import operator

from elements import ReplicaState
from elements.graph_element import GraphElement, GraphQuery, Op
from elements.element_feature.set_feature import PartSetElementFeature
from typing import TYPE_CHECKING, Tuple
from exceptions import OldVersionException, NotUsedOnReplicaException

if TYPE_CHECKING:
    from storage.process_fn import GraphStorageProcess


class ReplicableGraphElement(GraphElement):
    """ General Graph element that is replicatable. Used in Vertex for now can be added for other things as well """

    def __init__(self, master: int = None, is_halo=False, *args, **kwargs):
        super(ReplicableGraphElement, self).__init__(*args, **kwargs)
        self.master = master
        self._clock = 0
        self._halo = is_halo

    def __call__(self, rpc: "Rpc") -> bool:
        if self.state == ReplicaState.REPLICA:
            # Send this message to master node if it is replica
            query = GraphQuery(op=Op.RPC, element=rpc, part=self.master_part, iterate=True)
            self.storage.message(query)
            return False
        is_updated = super(ReplicableGraphElement, self).__call__(rpc)
        if is_updated:
            self.sync_replicas()
        return is_updated

    def create_element(self) -> bool:
        is_created = super(ReplicableGraphElement, self).create_element()  # Store
        if not is_created: return is_created
        if self.state is ReplicaState.MASTER:
            self['parts'] = PartSetElementFeature({self.storage.part_id})
        elif self.state is ReplicaState.REPLICA:
            query = GraphQuery(Op.SYNC, self, self.master_part, True)
            self.storage.message(query)
        return is_created

    def sync_element(self, new_element: "GraphElement") -> Tuple[bool, "GraphElement"]:
        if self.state is ReplicaState.MASTER:
            """ Add to parts and sync with replicas """
            self["parts"].add(new_element.part_id)  # This is going to handle syncing no need to bother
            return False, self
        elif self.state is ReplicaState.REPLICA:
            if new_element.integer_clock < self.integer_clock: raise OldVersionException
            return self.update_element(new_element)

    def external_update(self, new_element: "GraphElement") -> Tuple[bool, "GraphElement"]:
        if self.state is ReplicaState.REPLICA:
            query = GraphQuery(Op.UPDATE, new_element, self.master_part, True)
            self.storage.message(query)
            return False, self
        is_updated, memento = super(ReplicableGraphElement, self).external_update(new_element)  #Basically calling update_element
        if is_updated and self.state is ReplicaState.MASTER:
            self.sync_replicas()
        return is_updated, memento

    def __iter__(self):
        tmp = super(ReplicableGraphElement, self).__iter__()
        return filter(lambda x: x[0] != "parts", tmp)

    @property
    def master_part(self) -> int:
        return self.master

    def get_integer_clock(self):
        return self._clock

    def set_integer_clock(self, value: int):
        self._clock = value

    def del_integer_clock(self):
        del self._clock

    @property
    def is_halo(self) -> bool:
        return self._halo

    @property
    def replica_parts(self) -> list:
        re: "PartSetElementFeature" = self['parts']
        a = list(re.value)
        a.remove(self.part_id)
        return a

    @property
    def is_replicable(self) -> bool:
        return True

    def __getstate__(self):
        """ No need to serialize the parts """
        state = super(ReplicableGraphElement, self).__getstate__()
        if "parts" in state["_features"]:
            del state["_features"]["parts"]
        return state

File: src/main/java/elements/ReplicableGraphElement.java
package elements;

import features.Set;
import iterations.IterationType;
import iterations.Rmi;
import scala.Tuple2;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class ReplicableGraphElement extends GraphElement {
    public short master;
    public boolean halo;

    public ReplicableGraphElement() {
        super();
        this.master = -1;
        this.halo = false;
    }

    public ReplicableGraphElement(String id) {
        super(id);
        this.master = -1;
        this.halo = false;
    }

    public ReplicableGraphElement(String id, boolean halo) {
        super(id);
        this.halo = halo;
        this.master = -1;
    }

    public ReplicableGraphElement(String id, boolean halo, short master) {
        super(id);
        this.halo = halo;
        this.master = master;
    }

    @Override
    public GraphElement copy() {
        ReplicableGraphElement tmp = new ReplicableGraphElement(this.id, this.halo, this.master);
        tmp.partId = this.partId;
        return tmp;
    }

    @Override
    public GraphElement deepCopy() {
        ReplicableGraphElement tmp = new ReplicableGraphElement(this.id, this.halo, this.master);
        tmp.partId = this.partId;
        tmp.storage = this.storage;
        tmp.features.addAll(this.features);
        return tmp;
    }

    // Main Logical Stuff
    @Override
    public Boolean createElement() {
        if (this.state() == ReplicaState.REPLICA) {
            this.features.clear();
        }
        boolean is_created = super.createElement();
        if (is_created) {
            if (this.state() == ReplicaState.MASTER) {
                // Add setFeature
                this.setFeature("parts", new Set<Short>(new ArrayList<>(), true));
            } else {
                // Send Query
                this.storage.layerFunction.message(new GraphOp(Op.SYNC, this.masterPart(), this, IterationType.ITERATE));
            }
        }
        return is_created;
    }

    @Override
    public Tuple2<Boolean, GraphElement> syncElement(GraphElement newElement) {
        if (this.state() == ReplicaState.MASTER) {
            Set<Integer> tmp = (Set<Integer>) this.getFeature("parts");
            Rmi.call(tmp, "add", newElement.getPartId());
            this.syncReplica(newElement.getPartId());

        } else if (this.state() == ReplicaState.REPLICA) {
            return this.updateElement(newElement);
        }

        return super.syncElement(this);

    }

    @Override
    public Tuple2<Boolean, GraphElement> externalUpdate(GraphElement newElement) {
        if (this.state() == ReplicaState.MASTER) {
            Tuple2<Boolean, GraphElement> tmp = super.externalUpdate(newElement);
            if (tmp._1) this.syncReplicas(true);
            return tmp;
        } else if (this.state() == ReplicaState.REPLICA) {
            this.storage.layerFunction.message(new GraphOp(Op.COMMIT, this.masterPart(), newElement, IterationType.ITERATE));
            return new Tuple2<>(false, this);
        } else return super.externalUpdate(newElement);
    }


    public void syncReplicas(boolean skipHalo) {
        if ((this.state() != ReplicaState.MASTER) || (this.isHalo() && skipHalo) || this.replicaParts() == null || this.replicaParts().isEmpty())
            return;
        this.cacheFeatures();
        ReplicableGraphElement cpy = (ReplicableGraphElement) this.copy();
        for (Feature feature : this.features) {
            if (skipHalo && feature.isHalo()) continue;
            Feature tmp = (Feature) feature.copy();
            if (tmp.isHalo()) {
                tmp.value = null;
            }
            cpy.setFeature(feature.getFieldName(), tmp);
        }
        this.replicaParts().forEach(part_id -> this.storage.layerFunction.message(new GraphOp(Op.SYNC, part_id, cpy, IterationType.ITERATE)));
    }

    public void syncReplica(short part_id) {
        if (this.state() != ReplicaState.MASTER) return;
        this.cacheFeatures();
        ReplicableGraphElement cpy = (ReplicableGraphElement) this.copy();
        for (Feature feature : this.features) {
            Feature tmp = (Feature) feature.copy();
            if (tmp.isHalo()) {
                tmp.value = null;
            }
            cpy.setFeature(feature.getFieldName(), tmp);
        }

        this.storage.layerFunction.message(new GraphOp(Op.SYNC, part_id, cpy, IterationType.ITERATE));
    }

    @Override
    public short masterPart() {
        return this.master;
    }

    @Override
    public Boolean isHalo() {
        return this.halo;
    }

    @Override
    public List<Short> replicaParts() {
        Set<Short> parts = (Set<Short>) this.getFeature("parts");
        if (Objects.isNull(parts)) return super.replicaParts();
        else {
            return parts.getValue();
        }
    }

    @Override
    public Boolean isReplicable() {
        return true;
    }
}
